File: Python3/1033.py
Cmax,D,Davg,n = map(eval,input().split())
st = []
for _ in range(n):
    a,b = map(eval,input().split())
    st.append((a,b))
st.append((0,D))
st.sort(key=lambda x:x[1])
if st[0][1] != 0:
    print("The maximum travel distance = 0.00")
else:
    now = 0
    ans = 0
    nowTank=0
    MAX = Cmax*Davg
    while now < n:
        k = -1
        priceMin = 10000
        for i in range(now+1,n+1):
            if st[i][1] - st[now][1] > MAX:
                break
            if st[i][0] < priceMin:
                priceMin = st[i][0]
                k = i
                if priceMin < st[now][0]:
                    break
        if k == -1:
            break
        need = (st[k][1]-st[now][1])/Davg
        if priceMin < st[now][0]:
            if nowTank < need:
                ans += (need-nowTank)*st[now][0]
                nowTank = 0
            else:
                nowTank -= need
        else:
            ans += (Cmax-nowTank)*st[now][0]
            nowTank = Cmax-need
        now = k
    if now == n:
        print("%.2f"%ans)
    else:
        print("The maximum travel distance = %.2f"%(st[now][1]+MAX))

File: c++/1033.cpp
//
//  main.cpp
//  test
//
//  Created by 李寻欢 on 2019/10/4.
//  Copyright © 2019 李寻欢. All rights reserved.
//
#include <stdio.h>
#include <iostream>
#include <string>
#include <map>
#include <cmath>
#include <algorithm>
#include <stack>
#include <vector>
#include <queue>
#include <unordered_map>
#include <string.h>
#include <set>
using namespace std;


struct gas_station {
    double price;
    double distance;
}station[510];

bool cmp(gas_station a,gas_station b){
    if(a.distance != b.distance)
        return a.distance < b.distance;
    return a.price < b.price;
}

int main(){
    int tank,distance,unit,number;
    scanf("%d%d%d%d",&tank,&distance,&unit,&number);
    for (int i=0; i<number; i++) {
        scanf("%lf%lf",&station[i].price,&station[i].distance);
    }
    station[number].price = 0;
    station[number].distance = distance;
    sort(station, station+number, cmp);
    if(station[0].distance > 0){
        printf("The maximum travel distance = 0.00\n");
        return 0;
    }
    int now = 0;
    double ans = 0,nowtank=0,MAX = tank*unit;
    while (now < number) {
        int k = -1;
        double pricemin = 10000;
        for (int i=now+1; i<=number && station[i].distance - station[now].distance <= MAX; i++) {
            if (station[i].price < pricemin) {
                pricemin = station[i].price;
                k = i;
            }
            if (pricemin < station[now].price) {
                break;
            }
        }
        if (k == -1) {
            break;
        }
        double need = (station[k].distance - station[now].distance) / unit;
        if (pricemin < station[now].price) {
            if (nowtank < need) {
                ans += (need - nowtank)*station[now].price;
                nowtank = 0;
            }else{
                nowtank -= need;
            }
        }else{
            ans += (tank - nowtank)*station[now].price;
            nowtank = tank - need;
        }
        now = k;
    }
    if (now == number) {
        printf("%.2f\n",ans);
    }else
        printf("The maximum travel distance = %.2f\n",station[now].distance + MAX);
}


